### PR TITLE
REG: dont call func on empty input

### LINKED ISSUE #31760: groupby().agg() calls the user function for an extra time with empty inputs in 1.0.0
#### Code Sample, a copy-pastable example if possible

```python
df = pd.DataFrame({"key": ["a", "b", "c", "c"], "value": [1, 2, 3, 4]})                                                                                                                                                                                                                                                                                                                                                                                                                        

def foo(x): 
    print(f"called with {x}") 
    return len(x) 

df.groupby("key")["value"].agg(foo)
```
#### Problem description

In 1.0.0, this code outputs:
```
called with Series([], Name: value, dtype: int64)
called with 0    1
Name: value, dtype: int64
called with 1    2
Name: value, dtype: int64
called with 2    3
3    4
Name: value, dtype: int64
Out[10]: 
key
a    1
b    1
c    2
Name: value, dtype: int64
```

In 0.25.3, the code outputs:
```
In [5]:  df.groupby("key")["value"].agg(foo)                                                                                                                                                                                                                                                                                                                                                                                                                                                           
called with 0    1
Name: value, dtype: int64
called with 1    2
Name: value, dtype: int64
called with 2    3
3    4
Name: value, dtype: int64
Out[5]: 
key
a    1
b    1
c    2
Name: value, dtype: int64
```

In 1.0.0, `foo` is called one more time with empty input, which can break user code


**Note**: We receive a lot of issues on our GitHub tracker, so it is very possible that your issue has been posted before. Please check first before submitting so that we do not have to handle and close duplicates!

**Note**: Many problems can be resolved by simply upgrading `pandas` to the latest version. Before submitting, please check if that solution works for you. If possible, you may want to check if `master` addresses this issue, but that is not necessary.

For documentation-related issues, you can check the latest versions of the docs on `master` here:

https://pandas-docs.github.io/pandas-docs-travis/

If the issue has not been resolved there, go ahead and file it in the issue tracker.

#### Expected Output

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]

</details>

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #31760
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Could use confirmation from OP that this does in fact close #31760.

### CODE REVIEW

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 1]
```diff
@@ -13,6 +13,18 @@
 from pandas.core.groupby.grouper import Grouping
 
 
+def test_groupby_agg_no_extra_calls():
+    # GH#31760
+    df = pd.DataFrame({"key": ["a", "b", "c", "c"], "value": [1, 2, 3, 4]})
+    gb = df.groupby("key")["value"]
+
+    def dummy_func(x):
```
Comment (WillAyd) [inline pandas/tests/groupby/aggregate/test_aggregate.py]: I think a better approach here would be to Mock a function and examine the call_count after application. There's an example using `create_autospec` in the Python docs

https://docs.python.org/3/library/unittest.mock.html

File: pandas/core/groupby/groupby.py [hunk 2]
```diff
@@ -923,17 +923,10 @@ def _python_agg_general(self, func, *args, **kwargs):
 
             try:
                 # if this function is invalid for this dtype, we will ignore it.
-                func(obj[:0])
+                result, counts = self.grouper.agg_series(obj, f)
```
Comment (WillAyd) [inline pandas/core/groupby/groupby.py]: Any perf difference for numpy functions?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/aggregate/test_aggregate.py (hunk 1 above)
Comment (jreback) [reply pandas/tests/groupby/aggregate/test_aggregate.py]: we already use this pattern elsewhere

File: pandas/core/groupby/groupby.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/groupby/groupby.py]: i think this just means we dont fail-fast

**Review by WillAyd — APPROVED** [review]
lgtm

**Comment by jreback** [comment]
thanks @jbrockmendel

**Pull request merged**